File: app/securesocial/provider/providers/InstagramProvider.java
/**
 * Copyright 2011 Brian Porter (brian at porter dot net) - twitter: @poornerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.provider.providers;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import play.Logger;
import play.libs.WS;
import securesocial.provider.*;

import java.util.Map;

/**
 * A provider for Instagram 
 */
public class InstagramProvider extends OAuth2Provider {
    private static final String SELF_API = "https://api.instagram.com/v1/users/self/?access_token=%s";
    private static final String META = "meta";
    private static final String CODE = "code";
    private static final String ERROR_TYPE = "errorType";
    private static final String ERROR_DETAIL = "errorDetail";
    private static final String USER = "data";
    private static final String ID = "id";
    private static final String FULL_NAME = "full_name";
    private static final String SPACE = " ";
    private static final String PHOTO = "profile_picture";
    private static final String CONTACT = "contact";
    private static final String EMAIL = "email";

    public InstagramProvider() {
        super(ProviderType.instagram);
    }

    @Override
    protected void fillProfile(SocialUser user, Map<String, Object> authContext) {
        JsonObject me = WS.url(SELF_API, user.accessToken).get().getJson().getAsJsonObject();
        JsonObject meta = me.getAsJsonObject(META);

        if ( meta.get(CODE).getAsInt() != 200 ) {
            final String errorType = meta.get(ERROR_TYPE).getAsString();
            final String errorDetail = meta.get(ERROR_DETAIL).getAsString();
            Logger.error("Error retrieving profile information from Instagram. Error type: %s, detail: %s.", errorType, errorDetail);
            throw new AuthenticationException();
        }


        JsonObject userInfo = me.getAsJsonObject(USER);
        if( userInfo == null ) {
            final String errorType = userInfo.get(ERROR_TYPE).getAsString();
            final String errorDetail = userInfo.get(ERROR_DETAIL).getAsString();
            Logger.error("Error retrieving profile data from Instagram. Error type: %s, detail: %s.", errorType, errorDetail);
            throw new AuthenticationException();
        }

        user.id.id = userInfo.get(ID).getAsString();
        user.displayName = userInfo.get(FULL_NAME).getAsString();
        user.avatarUrl = userInfo.get(PHOTO).getAsString();
        final JsonObject contact = userInfo.getAsJsonObject(CONTACT);
        if ( contact != null ) {
            final JsonElement userEmail = contact.get(EMAIL);
            if ( userEmail != null ) {
                user.email = userEmail.getAsString();
            }
        }
    }

}

File: module-code/app/securesocial/core/providers/InstagramProvider.scala
/**
 * Copyright 2013 Brian Porter (poornerd at gmail dot com) - twitter: @poornerd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.core.providers

import securesocial.core._
import securesocial.core.services.{CacheService, RoutesService}

import scala.concurrent.Future

/**
 * An Instagram provider
 *
 */
class InstagramProvider(routesService: RoutesService,
                        cacheService: CacheService,
                        client: OAuth2Client)
  extends OAuth2Provider(routesService, client, cacheService)
{
  val GetAuthenticatedUser = "https://api.instagram.com/v1/users/self?access_token=%s"
  val AccessToken = "access_token"
  val TokenType = "token_type"
  val Data = "data"
  val Username = "username"
  val FullName ="full_name"
  val ProfilePic = "profile_picture"
  val Id = "id"
  
  override val id = InstagramProvider.Instagram

  def fillProfile(info: OAuth2Info): Future[BasicProfile] = {
    import scala.concurrent.ExecutionContext.Implicits.global
    client.retrieveProfile(GetAuthenticatedUser.format(info.accessToken)).map { me =>
        (me \ "response" \ "user").asOpt[String] match {
          case Some(msg) => {
            logger.error(s"[securesocial] error retrieving profile information from Instagram. Message = $msg")
            throw new AuthenticationException()
          }
          case _ =>
            val userId = (me \ Data \ Id).as[String]
            val fullName = (me \ Data \ FullName).asOpt[String]
            val avatarUrl = (me \ Data \ ProfilePic).asOpt[String]
            BasicProfile(id, userId, None, None, fullName, None, avatarUrl, authMethod, oAuth2Info = Some(info))
        }
    } recover {
      case e: AuthenticationException => throw e
      case e: Exception =>
        logger.error( "[securesocial] error retrieving profile information from Instagram", e)
        throw new AuthenticationException()
    }
  }
}

object InstagramProvider {
  val Instagram = "instagram"
}
